dev/README.md

```diff
@@ -0,0 +1,666 @@
+The unit tests are mostly written with in-memory implementations of various external resources,
+most notable one is the Kubernetes client where we use [k8s.io/client-go/kubernetes/fake](https://pkg.go.dev/k8s.io/client-go/kubernetes/fake) to create an in-memory Kubernetes client and add objects to it for testing.
+
+The unit tests are scattered in `/cmd` and `/pkg` folders in `*_test.go` files.
+
+Running unit tests requires you to have Go installed, you can follow [Go's installation guidance](https://go.dev/doc/install) to install it.
+
+These tests runs usually pretty fast, and you can run all test suites using:
+```bash
+$ pwd
+~/mountpoint-s3-csi-driver
+
+$ make test
+...
+ok      github.com/awslabs/mountpoint-s3-csi-driver/tests/sanity        0.657s
+```
+
+This command runs both unit tests and also [sanity test of Kubernetes CSI](https://kubernetes-csi.github.io/docs/unit-testing.html).
+
+You can also use Go directly run specific tests if you're working on a specific feature, for example:
+```bash
+# Only run credential provider tests
+$ go test ./pkg/driver/node/credentialprovider/...
+ok      github.com/awslabs/mountpoint-s3-csi-driver/pkg/driver/node/credentialprovider  1.004s
+ok      github.com/awslabs/mountpoint-s3-csi-driver/pkg/driver/node/credentialprovider/awsprofile       0.283s
+?       github.com/awslabs/mountpoint-s3-csi-driver/pkg/driver/node/credentialprovider/awsprofile/awsprofiletest        [no test files]
+?       github.com/awslabs/mountpoint-s3-csi-driver/pkg/driver/node/credentialprovider/mocks    [no test files]
+
+# Only run Mountpoint Pod Watcher tests
+$ go test -v ./pkg/podmounter/mppod/watcher/watcher_test.go
+=== RUN   TestGettingAlreadyScheduledAndReadyPod
+--- PASS: TestGettingAlreadyScheduledAndReadyPod (0.15s)
+=== RUN   TestGettingScheduledButNotYetReadyPod
+--- PASS: TestGettingScheduledButNotYetReadyPod (0.11s)
+=== RUN   TestGettingNotYetScheduledPod
+--- PASS: TestGettingNotYetScheduledPod (0.13s)
+=== RUN   TestGettingUnschedulablePod
+--- PASS: TestGettingUnschedulablePod (0.12s)
+=== RUN   TestGettingPodsConcurrently
+--- PASS: TestGettingPodsConcurrently (0.11s)
+PASS
+ok      command-line-arguments  1.159s
+```
+
+#### Controller tests
+
+The [controller component](/docs/ARCHITECTURE.md#the-controller-component-aws-s3-csi-controller) of the CSI Driver
+is mainly responsible for managing Mountpoint Pods by communicating with the Kubernetes control plane.
+
+The CSI Driver uses [envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest) to setup a local Kubernetes control plane and run test against it.
+
+These controller tests can be found in `/tests/controller` folder, and can be run by:
+
+```bash
+$ pwd
+~/mountpoint-s3-csi-driver
+
+$ make test
+...
+Ran 55 of 55 Specs in 159.773 seconds
+SUCCESS! -- 55 Passed | 0 Failed | 0 Pending | 0 Skipped
+--- PASS: TestControllers (159.77s)
+PASS
+ok      github.com/awslabs/mountpoint-s3-csi-driver/tests/controller    160.335s
+```
```

Comment:
I remember you saying these might be incorrect so leaving this comment as a reminder